File: surface/widgets/swiftui/components/widgets/domain/CanvasNode.swift
// ============================================================
// Clef Surface SwiftUI Widget — CanvasNode
//
// Individual element on a canvas surface rendered as a draggable
// card. Supports selection, label display, position info, and
// drag interaction for repositioning on the canvas.
// ============================================================

import SwiftUI

struct CanvasPosition {
    let x: CGFloat
    let y: CGFloat
}

struct CanvasNodeView: View {
    var id: String
    var label: String
    var position: CanvasPosition = CanvasPosition(x: 0, y: 0)
    var selected: Bool = false
    var type: String? = nil
    var onSelect: (String) -> Void = { _ in }
    var onDrag: (String, CGSize) -> Void = { _, _ in }

    @GestureState private var dragOffset: CGSize = .zero

    var body: some View {
        VStack(alignment: .leading, spacing: 4) {
            Text(label + (type != nil ? " [\(type!)]" : ""))
                .font(.subheadline)
                .fontWeight(selected ? .bold : .regular)
                .foregroundColor(selected ? .accentColor : .primary)

            Text("(\(Int(position.x)), \(Int(position.y)))" + (selected ? " [selected]" : ""))
                .font(.caption)
                .foregroundColor(.secondary)
        }
        .padding(12)
        .background(
            RoundedRectangle(cornerRadius: 8)
                .fill(selected ? Color.accentColor.opacity(0.1) : Color(.systemBackground))
        )
        .overlay(
            RoundedRectangle(cornerRadius: 8)
                .stroke(selected ? Color.accentColor : Color(.systemGray3), lineWidth: selected ? 2 : 1)
        )
        .onTapGesture { onSelect(id) }
        .gesture(
            DragGesture()
                .updating($dragOffset) { value, state, _ in
                    state = value.translation
                }
                .onEnded { value in
                    onDrag(id, value.translation)
                }
        )
        .accessibilityLabel("Canvas node: \(label)")
        .accessibilityHint(selected ? "Selected" : "Tap to select")
    }
}

File: surface/widgets/nativescript/components/widgets/domain/CanvasNode.ts
// ============================================================
// Clef Surface NativeScript Widget — CanvasNode
//
// Draggable node on a canvas with ports and content.
// ============================================================

import { StackLayout, Label } from '@nativescript/core';
import type { View } from '@nativescript/core';

export interface CanvasNodeProps {
  id?: string;
  x?: number;
  y?: number;
  width?: number;
  height?: number;
  title?: string;
  selected?: boolean;
  onMove?: (x: number, y: number) => void;
  onSelect?: () => void;
  children?: View[];
}

export function createCanvasNode(props: CanvasNodeProps): StackLayout {
  const { id, x = 0, y = 0, width = 200, height, title = 'Node', selected = false, onMove, onSelect, children = [] } = props;
  const container = new StackLayout();
  container.className = `clef-widget-canvas-node${selected ? ' clef-selected' : ''}`;
  container.width = width;
  if (height) container.height = height;
  const titleLabel = new Label();
  titleLabel.text = title;
  titleLabel.fontWeight = 'bold';
  titleLabel.padding = '4 8';
  container.addChild(titleLabel);
  for (const child of children) container.addChild(child);
  container.on('tap', () => onSelect?.());
  return container;
}

export default createCanvasNode;
